dcc-storage-client/src/main/java/org/icgc/dcc/storage/client/upload/s3/S3UploadService.java

```diff
@@ -0,0 +1,240 @@
+    ObjectSpecification spec = null;
+    try {
+      spec = storageService.initiateUpload(objectId, file.length(), overwrite, md5);
+    } catch (NotRetryableException e) {
+      // A NotRetryable exception during initiateUpload should just end whole process
+      // a bit of a sleazy hack. Should only be thrown when the Metadata service informs us the supplied
+      // object id was never registered/does not exist in Metadata repo
+      throw new NotResumableException(e);
+    }
+
+    // Delete if already present
+    if (overwrite) {
+      UploadStateStore.create(file, spec, false);
+    }
+
+    val progress = new Progress(terminal, quiet, spec.getParts().size(), 0);
+    uploadParts(spec.getParts(), file, objectId, spec.getUploadId(), progress);
+    cleanupState(file, objectId);
+  }
+
+  /**
+   * Resume a upload if it is possible. Otherwise, it will start a new upload. Resume might not be possible if the
+   * upload progress cannot be retrieved.
+   */
+  @SneakyThrows
+  private void resumeIfPossible(File uploadFile, String objectId, String md5, boolean checksum) {
+    UploadProgress progress = null;
+    try {
+      progress = checkProgress(uploadFile, objectId);
+    } catch (NotRetryableException e) {
+      // org.icgc.dcc.storage.client.exception.ServiceRetryableResponseErrorHandler translates the 404 received from
+      // server into a NotRetryableException
+      log.info("No upload id found for object id {}. Start new upload.", objectId);
+      startUpload(uploadFile, objectId, md5, true);
+      return;
+    }
+    resume(uploadFile, progress, objectId, checksum);
+  }
+
+  @SneakyThrows
+  private UploadProgress checkProgress(File uploadFile, String objectId) {
+
+    // See if there is already an upload in progress for this object id. Fetch upload id and send if present. If
+    // missing, send null
+    val uploadId = UploadStateStore.fetchUploadId(uploadFile, objectId);
+    UploadProgress progress = null;
+    progress = storageService.getProgress(objectId, uploadFile.length());
+
+    // Compare upload id's
+    if (uploadId.isPresent() && progress != null) {
+      if (!uploadId.get().equalsIgnoreCase(progress.getUploadId())) {
+        // Can't continue; upload id's don't match
+        val msg =
+            String.format(
+                "Local in-progress upload %s conflicts with remote upload in progress %s. Aborting local upload.",
+                uploadId.get(),
+                progress.getUploadId());
+        throw new NotResumableException(new IllegalStateException(msg));
+      }
+      // Then local upload id not present
+    } else if (progress != null) {
+      val msg =
+          String
+              .format(
+                  "There is already an upload in progress that was started elsewhere with upload id %s. Aborting local upload.",
+                  progress.getUploadId());
+      throw new NotResumableException(new IllegalStateException(msg));
```

Comment:
This pattern happens very very often throughout the code base and takes significant amount of space to articulate. I think adding a `checkStateResumable` might be in order. For example:

```java
...
} else {
  checkStateResumable(progress == null, "There is already an upload in progress that was started elsewhere   with upload id %s. Aborting local upload.",  
  progress.getUploadId());
}
```